Plugins/BuildServerIntegration/AzureDevOpsIntegration/AzureDevOpsAdapter.cs

```diff
@@ -144,97 +145,132 @@ private async Task ObserveBuildsAsync(DateTime? sinceDate, bool running, IObserv
+            }
+        }
 
-                void ManageError(string errorMessage)
-                {
-                    _apiClient = null;
-                    observer.OnNext(new BuildInfo { CommitHashList = new[] { ObjectId.WorkTreeId }, Description = errorMessage, Status = BuildInfo.BuildStatus.Failure });
-                    observer.OnCompleted();
-                }
+        private async Task<bool> EnsureBuildDefinitionsIsInitializedAsync(IObserver<BuildInfo> observer)
+        {
+            if (_buildDefinitions is not null)
+            {
+                return true;
+            }
+
+            try
+            {
+                Validates.NotNull(_buildDefinitionsTask);
+
+                _buildDefinitions = await _buildDefinitionsTask.JoinAsync();
 
                 if (_buildDefinitions is null)
                 {
                     observer.OnCompleted();
-                    return;
+                    return false;
                 }
 
-                var builds = running ?
-                    FilterRunningBuilds(await _apiClient.QueryRunningBuildsAsync(_buildDefinitions)) :
-                    await _apiClient.QueryFinishedBuildsAsync(_buildDefinitions, sinceDate);
+                _buildsCache = new BuildsCache { Id = CacheKey, BuildDefinitions = _buildDefinitions };
+                return true;
+            }
+            catch (UnauthorizedAccessException)
+            {
+                string errorMessage = _badTokenErrorMessage.Text;
+                ManageError(errorMessage);
 
-                foreach (var build in builds)
+                if (_projectOnErrorKey is null || _projectOnErrorKey != CacheKey)
                 {
-                    observer.OnNext(CreateBuildInfo(build));
+                    _projectOnErrorKey = CacheKey;
+
+                    TaskDialogButton btnOpenSettings = new("Open settings");
+                    TaskDialogButton btnIgnore = new("Ignore");
+                    TaskDialogPage page = new()
+                    {
+                        Heading = errorMessage,
+                        Icon = TaskDialogIcon.Error,
+                        AllowCancel = true,
+                        Caption = _buildIntegrationErrorCaption.Text,
+                        Buttons = { btnOpenSettings, btnIgnore }
+                    };
+
+                    TaskDialogButton result = TaskDialog.ShowDialog(page);
+                    if (result == btnOpenSettings)
+                    {
+                        _projectOnErrorKey = null;
+                        Validates.NotNull(_openSettings);
+                        _openSettings.Invoke();
+                    }
                 }
+
+                return false;
             }
-            catch (OperationCanceledException)
+            catch (Exception ex)
             {
-                // Do nothing, the observer is already stopped
+                string errorMessage = _genericErrorMessage.Text + ex.Message;
```

Comment:
String interpolation is getting perf optimisations, so it's better to do this:
```suggestion
                string errorMessage = $"{_genericErrorMessage.Text}{ex.Message}";
```

Shouldn't there be some sort of a separator? E.g., like a new line or something like that...